Make uninit checks stricter but avoid issues for old hyper

5225225 opened the pull request:
This makes the checks for `std::mem::uninitialized` stricter, but tries to avoid making `hyper` panic.

`hyper` versions 0.11 to early 0.14 all make slices inside arrays, so we don't panic on slices inside arrays. We still will panic on any other reference, even to ZSTs.

No change was made to the `std::mem::zeroed` checks.

We still consider uninit primitives that have a full range to be valid, as well as &[u8] (more generally, any type with an align of 1) and &str but *only* inside arrays.

I ran these on some known bad `hyper` crates, and they all passed, whereas a known bad crossbeam crate using mem::zeroed failed. Which is what I expected.

See: #66151

Once the code's been reviewed and we're happy with the amount of checks this is doing, this will definitely need a crater run. I'm expecting hopefully less than the previous run, mostly crossbeam but maybe with some extra surprises.

Commit:
Make uninit checks stricter but avoid issues for old hyper

rust-highfive commented:
r? @oli-obk

(rust-highfive has picked a reviewer for you, use r? to override)

oli-obk commented:
@bors try

bors commented:
:hourglass: Trying commit 168fa786e9c38469881a8490b7752e75de90f1b1 with merge af445596a53bb2f55a69322f552f9f231e2522ad...
<!-- homu: {"type":"TryBuildStarted","head_sha":"168fa786e9c38469881a8490b7752e75de90f1b1","merge_sha":"af445596a53bb2f55a69322f552f9f231e2522ad"} -->

5225225 commented:
Also, this is strictly speaking only not LLVM ub for mem::uninit if/once #99182 gets merged, but no need to block this on that getting in, it's *currently* LLVM ub for hyper to do what its doing.

Also gonna cc @RalfJung if you want to have a look at this. I did mostly follow what you described, but also figured the const eval checks for mem::zeroed would be fine since not many people misuse that, I don't expect us to be breaking a lot more code since the only thing that breaks is someone doing `Result<NonZeroT, _>` or any other enum where the 0 variant doesn't allow being left zeroed.

Plus it exercises the const eval check code path by default, so no surprises if/when we go to turn it on everywhere. Which is probably a *long* way off considering how much code makes uninit ints, but I'm optimistic we can do it Eventually.

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/7386474368?check_suite_focus=true)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/7386474368?check_suite_focus=true","builder_name":"checks-actions"} -->

oli-obk commented on compiler/rustc_middle/src/ty/layout.rs:
```diff
@@ -3482,3 +3482,86 @@ fn make_thin_self_ptr<'tcx>(
+///
+/// This code is intentionally conservative, and will not detect
+/// * making uninitialized types who have a full valid range (ints, floats, raw pointers)
+/// * uninit invalid &[T] where T has align 1 (only inside arrays)
```
```
error: unresolved link to `T`
--> compiler/rustc_middle/src/ty/layout.rs:3491:24
|
3491|/// * uninit invalid &[T] where T has align 1 (only inside arrays)
| ^no item named `T` in scope
|
= note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
= help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
```

I guess you need to put these into backticks ^^

5225225 replied:
Right. Yeah, that's the second time I messed up a docs thing, I really gotta get into the habit of checking x.py doc :)

Will fix now, sorry about that :)

(How expensive would it be for PR CI to run docs? I don't know if I'm just the only one hitting this or if it would be worth adding it to PR CI.)

5225225 replied:
```
workspace: /home/jess/src/rust/Cargo.toml
   Compiling serde v1.0.125
   Compiling rand v0.8.5
   Compiling tempfile v3.2.0
   Compiling serde_json v1.0.59
   Compiling lint-docs v0.1.0 (/home/jess/src/rust/src/tools/lint-docs)
    Finished release [optimized] target(s) in 6.08s
warning: the code example in lint `unfulfilled_lint_expectations` in /home/jess/src/rust/compiler/rustc_lint_defs/src/builtin.rs failed to generate the expected output: did not find lint `unfulfilled_lint_expectations` in output of example, got:

error[E0554]: `#![feature]` may not be used on the beta release channel
 --> lint_example.rs:1:1
  |
1 | #![feature(lint_reasons)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^


error: aborting due to previous error


For more information about this error, try `rustc --explain E0554`.

warning: the code example in lint `unused_allocation` in /home/jess/src/rust/compiler/rustc_lint/src/unused.rs failed to generate the expected output: did not find lint `unused_allocation` in output of example, got:

error[E0554]: `#![feature]` may not be used on the beta release channel
 --> lint_example.rs:1:1
  |
1 | #![feature(box_syntax)]
  | ^^^^^^^^^^^^^^^^^^^^^^^


error: aborting due to previous error


For more information about this error, try `rustc --explain E0554`.

Rustbook (x86_64-unknown-linux-gnu) - rustc
Rustbook (x86_64-unknown-linux-gnu) - cargo
```

known issue? I saw that on master doc runs too. In any case, I couldn't see any failures that seemed to be caused by me, so.... It's probably fine now.

jyn514 replied:
> How expensive would it be for PR CI to run docs?

Not too terribly long, but for it to be reliable we can't use stage 0 rustdoc, which means we have to add it to the LLVM-12 builder, which is already kinda slow :/

jyn514 replied:
Oh also do you mind opening an issue for that feature gate error? We should fix it, but we *really* should make sure it's tested somewhere and not just ignored.

Commit:
Fix docs

rust-log-analyzer commented:
The job **`dist-x86_64-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/7385313446?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/af445596a53bb2f55a69322f552f9f231e2522ad/checks/7385313446/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Preparing bitmaps-3.1.0
[2022-07-18T08:47:32Z INFO  collector::execute] run_rustc with incremental=false, profile=Debug, scenario=None, patch=None
[2022-07-18T08:47:32Z INFO  collector::execute] run_rustc with incremental=false, profile=Check, scenario=None, patch=None
[2022-07-18T08:47:32Z INFO  collector::execute] run_rustc with incremental=false, profile=Opt, scenario=None, patch=None
[2022-07-18T08:47:32Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpbvkHoO#bitmaps@3.1.0" "--profile" "check" "--" "--skip-this-rustc"
[2022-07-18T08:47:32Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp9oVLKt#bitmaps@3.1.0" "--" "--skip-this-rustc"
[2022-07-18T08:47:32Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpxLi4pg#bitmaps@3.1.0" "--release" "--" "--skip-this-rustc"
[2022-07-18T08:47:33Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-07-18T08:47:33Z INFO  collector::execute] run_rustc with incremental=false, profile=Check, scenario=Some(Full), patch=None
[2022-07-18T08:47:33Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpjxegD8#bitmaps@3.1.0" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln"
[2022-07-18T08:47:35Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrFull), patch=None
[2022-07-18T08:47:35Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrFull), patch=None
[2022-07-18T08:47:35Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpjxegD8#bitmaps@3.1.0" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpjxegD8/incremental-state"
[2022-07-18T08:47:37Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrUnchanged), patch=None
[2022-07-18T08:47:37Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpjxegD8#bitmaps@3.1.0" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpjxegD8/incremental-state"
[2022-07-18T08:47:38Z DEBUG collector::execute] applying println to "/tmp/.tmpjxegD8"
[2022-07-18T08:47:38Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2022-07-18T08:47:38Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2022-07-18T08:47:38Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpjxegD8#bitmaps@3.1.0" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpjxegD8/incremental-state"
[2022-07-18T08:47:39Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-07-18T08:47:39Z INFO  collector::execute] run_rustc with incremental=false, profile=Debug, scenario=Some(Full), patch=None
[2022-07-18T08:47:39Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpIweT2g#bitmaps@3.1.0" "--" "--wrap-rustc-with" "Eprintln"
[2022-07-18T08:47:41Z INFO  collector::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrFull), patch=None
---
[2022-07-18T08:47:46Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-07-18T08:47:46Z INFO  collector::execute] run_rustc with incremental=false, profile=Opt, scenario=Some(Full), patch=None
[2022-07-18T08:47:46Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpVxHvRn#bitmaps@3.1.0" "--release" "--" "--wrap-rustc-with" "Eprintln"
[2022-07-18T08:47:48Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrFull), patch=None
[2022-07-18T08:47:48Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpVxHvRn#bitmaps@3.1.0" "--release" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpVxHvRn/incremental-state"
[2022-07-18T08:47:52Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrUnchanged), patch=None
[2022-07-18T08:47:52Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpVxHvRn#bitmaps@3.1.0" "--release" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpVxHvRn/incremental-state"
[2022-07-18T08:47:52Z DEBUG collector::execute] applying println to "/tmp/.tmpVxHvRn"
[2022-07-18T08:47:52Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2022-07-18T08:47:52Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2022-07-18T08:47:52Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpVxHvRn#bitmaps@3.1.0" "--release" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpVxHvRn/incremental-state"
Executing benchmark cargo-0.60.0 (2/8)
Preparing cargo-0.60.0
[2022-07-18T08:47:53Z INFO  collector::execute] run_rustc with incremental=false, profile=Opt, scenario=None, patch=None
[2022-07-18T08:47:53Z INFO  collector::execute] run_rustc with incremental=false, profile=Debug, scenario=None, patch=None
---
[2022-07-18T08:52:14Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-07-18T08:52:14Z INFO  collector::execute] run_rustc with incremental=false, profile=Opt, scenario=Some(Full), patch=None
[2022-07-18T08:52:14Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp2gpRWm#cargo@0.60.0" "--release" "--lib" "--" "--wrap-rustc-with" "Eprintln"
[2022-07-18T08:53:14Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrFull), patch=None
[2022-07-18T08:53:14Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp2gpRWm#cargo@0.60.0" "--release" "--lib" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmp2gpRWm/incremental-state"
[2022-07-18T08:54:15Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrUnchanged), patch=None
[2022-07-18T08:54:15Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp2gpRWm#cargo@0.60.0" "--release" "--lib" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmp2gpRWm/incremental-state"
[2022-07-18T08:54:25Z DEBUG collector::execute] applying println to "/tmp/.tmp2gpRWm"
[2022-07-18T08:54:25Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2022-07-18T08:54:25Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2022-07-18T08:54:25Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp2gpRWm#cargo@0.60.0" "--release" "--lib" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmp2gpRWm/incremental-state"
Executing benchmark ctfe-stress-5 (3/8)
Preparing ctfe-stress-5
[2022-07-18T08:54:53Z INFO  collector::execute] run_rustc with incremental=false, profile=Check, scenario=None, patch=None
[2022-07-18T08:54:53Z INFO  collector::execute] run_rustc with incremental=false, profile=Debug, scenario=None, patch=None
---
[2022-07-18T08:55:32Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-07-18T08:55:32Z INFO  collector::execute] run_rustc with incremental=false, profile=Opt, scenario=Some(Full), patch=None
[2022-07-18T08:55:32Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpTFPoS5#ctfe-stress-5@0.1.0" "--release" "--" "--wrap-rustc-with" "Eprintln"
[2022-07-18T08:55:40Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrFull), patch=None
[2022-07-18T08:55:40Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpTFPoS5#ctfe-stress-5@0.1.0" "--release" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpTFPoS5/incremental-state"
[2022-07-18T08:55:51Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrUnchanged), patch=None
[2022-07-18T08:55:51Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpTFPoS5#ctfe-stress-5@0.1.0" "--release" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpTFPoS5/incremental-state"
Executing benchmark diesel-1.4.8 (4/8)
Preparing diesel-1.4.8
[2022-07-18T08:55:51Z INFO  collector::execute] run_rustc with incremental=false, profile=Check, scenario=None, patch=None
[2022-07-18T08:55:51Z INFO  collector::execute] run_rustc with incremental=false, profile=Opt, scenario=None, patch=None
---
[2022-07-18T08:56:10Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-07-18T08:56:10Z INFO  collector::execute] run_rustc with incremental=false, profile=Check, scenario=Some(Full), patch=None
[2022-07-18T08:56:10Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpcSs0Oh#diesel@1.4.8" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln"
[2022-07-18T08:56:27Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrFull), patch=None
[2022-07-18T08:56:27Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpcSs0Oh#diesel@1.4.8" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpcSs0Oh/incremental-state"
[2022-07-18T08:56:46Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrUnchanged), patch=None
[2022-07-18T08:56:46Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpcSs0Oh#diesel@1.4.8" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpcSs0Oh/incremental-state"
[2022-07-18T08:56:49Z DEBUG collector::execute] applying println to "/tmp/.tmpcSs0Oh"
[2022-07-18T08:56:49Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2022-07-18T08:56:49Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2022-07-18T08:56:49Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpcSs0Oh#diesel@1.4.8" "--profile" "check" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpcSs0Oh/incremental-state"
[2022-07-18T08:56:53Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-07-18T08:56:53Z INFO  collector::execute] run_rustc with incremental=false, profile=Debug, scenario=Some(Full), patch=None
[2022-07-18T08:56:53Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp9HtaxS#diesel@1.4.8" "--" "--wrap-rustc-with" "Eprintln"
[2022-07-18T08:57:11Z INFO  collector::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrFull), patch=None
[2022-07-18T08:57:11Z INFO  collector::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrFull), patch=None
[2022-07-18T08:57:11Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp9HtaxS#diesel@1.4.8" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmp9HtaxS/incremental-state"
[2022-07-18T08:57:33Z INFO  collector::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrUnchanged), patch=None
[2022-07-18T08:57:33Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp9HtaxS#diesel@1.4.8" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmp9HtaxS/incremental-state"
[2022-07-18T08:57:37Z DEBUG collector::execute] applying println to "/tmp/.tmp9HtaxS"
[2022-07-18T08:57:37Z INFO  collector::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2022-07-18T08:57:37Z INFO  collector::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("println"), path: "0-println.patch" })
[2022-07-18T08:57:37Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp9HtaxS#diesel@1.4.8" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmp9HtaxS/incremental-state"
[2022-07-18T08:57:41Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-07-18T08:57:41Z INFO  collector::execute] run_rustc with incremental=false, profile=Opt, scenario=Some(Full), patch=None
[2022-07-18T08:57:41Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp0mniMM#diesel@1.4.8" "--release" "--" "--wrap-rustc-with" "Eprintln"
[2022-07-18T08:58:00Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrFull), patch=None
---
[2022-07-18T08:58:33Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-07-18T08:58:33Z INFO  collector::execute] run_rustc with incremental=false, profile=Debug, scenario=Some(Full), patch=None
[2022-07-18T08:58:33Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpOBDOuh#externs@0.1.0" "--" "--wrap-rustc-with" "Eprintln"
[2022-07-18T08:58:34Z INFO  collector::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrFull), patch=None
[2022-07-18T08:58:34Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpOBDOuh#externs@0.1.0" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpOBDOuh/incremental-state"
[2022-07-18T08:58:35Z INFO  collector::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrUnchanged), patch=None
[2022-07-18T08:58:35Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpOBDOuh#externs@0.1.0" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpOBDOuh/incremental-state"
[2022-07-18T08:58:36Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-07-18T08:58:36Z INFO  collector::execute] run_rustc with incremental=false, profile=Opt, scenario=Some(Full), patch=None
[2022-07-18T08:58:36Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp9fjdrc#externs@0.1.0" "--release" "--" "--wrap-rustc-with" "Eprintln"
[2022-07-18T08:58:36Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrFull), patch=None
[2022-07-18T08:58:36Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrFull), patch=None
[2022-07-18T08:58:36Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp9fjdrc#externs@0.1.0" "--release" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmp9fjdrc/incremental-state"
[2022-07-18T08:58:38Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrUnchanged), patch=None
[2022-07-18T08:58:38Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp9fjdrc#externs@0.1.0" "--release" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmp9fjdrc/incremental-state"
Executing benchmark match-stress (6/8)
Preparing match-stress
[2022-07-18T08:58:38Z INFO  collector::execute] run_rustc with incremental=false, profile=Opt, scenario=None, patch=None
[2022-07-18T08:58:38Z INFO  collector::execute] run_rustc with incremental=false, profile=Debug, scenario=None, patch=None
---
[2022-07-18T08:58:46Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-07-18T08:58:46Z INFO  collector::execute] run_rustc with incremental=false, profile=Debug, scenario=Some(Full), patch=None
[2022-07-18T08:58:46Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmphIKhV4#match-stress@0.1.0" "--" "--wrap-rustc-with" "Eprintln"
[2022-07-18T08:58:49Z INFO  collector::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrFull), patch=None
[2022-07-18T08:58:49Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmphIKhV4#match-stress@0.1.0" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmphIKhV4/incremental-state"
[2022-07-18T08:58:52Z INFO  collector::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrUnchanged), patch=None
[2022-07-18T08:58:53Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmphIKhV4#match-stress@0.1.0" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmphIKhV4/incremental-state"
[2022-07-18T08:58:54Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-07-18T08:58:54Z INFO  collector::execute] run_rustc with incremental=false, profile=Opt, scenario=Some(Full), patch=None
[2022-07-18T08:58:54Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpDgk10O#match-stress@0.1.0" "--release" "--" "--wrap-rustc-with" "Eprintln"
[2022-07-18T08:58:57Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrFull), patch=None
---
[2022-07-18T08:59:03Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-07-18T08:59:03Z INFO  collector::execute] run_rustc with incremental=false, profile=Check, scenario=Some(Full), patch=None
[2022-07-18T08:59:03Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp6hkdaP#token-stream-stress@0.0.0" "--profile" "check" "--bin" "token-stream-stress-bin" "--" "--wrap-rustc-with" "Eprintln"
[2022-07-18T08:59:03Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrFull), patch=None
[2022-07-18T08:59:03Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp6hkdaP#token-stream-stress@0.0.0" "--profile" "check" "--bin" "token-stream-stress-bin" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmp6hkdaP/incremental-state"
[2022-07-18T08:59:03Z INFO  collector::execute] run_rustc with incremental=true, profile=Check, scenario=Some(IncrUnchanged), patch=None
[2022-07-18T08:59:03Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp6hkdaP#token-stream-stress@0.0.0" "--profile" "check" "--bin" "token-stream-stress-bin" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmp6hkdaP/incremental-state"
[2022-07-18T08:59:03Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-07-18T08:59:03Z INFO  collector::execute] run_rustc with incremental=false, profile=Debug, scenario=Some(Full), patch=None
[2022-07-18T08:59:03Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp32P3H6#token-stream-stress@0.0.0" "--bin" "token-stream-stress-bin" "--" "--wrap-rustc-with" "Eprintln"
[2022-07-18T08:59:03Z INFO  collector::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrFull), patch=None
---
[2022-07-18T08:59:04Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-07-18T08:59:04Z INFO  collector::execute] run_rustc with incremental=false, profile=Opt, scenario=Some(Full), patch=None
[2022-07-18T08:59:04Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpEuocBm#token-stream-stress@0.0.0" "--release" "--bin" "token-stream-stress-bin" "--" "--wrap-rustc-with" "Eprintln"
[2022-07-18T08:59:04Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrFull), patch=None
[2022-07-18T08:59:04Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpEuocBm#token-stream-stress@0.0.0" "--release" "--bin" "token-stream-stress-bin" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpEuocBm/incremental-state"
[2022-07-18T08:59:05Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrUnchanged), patch=None
[2022-07-18T08:59:05Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpEuocBm#token-stream-stress@0.0.0" "--release" "--bin" "token-stream-stress-bin" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpEuocBm/incremental-state"
Executing benchmark tuple-stress (8/8)
Preparing tuple-stress
[2022-07-18T08:59:05Z INFO  collector::execute] run_rustc with incremental=false, profile=Check, scenario=None, patch=None
[2022-07-18T08:59:05Z INFO  collector::execute] run_rustc with incremental=false, profile=Debug, scenario=None, patch=None
---
[2022-07-18T08:59:24Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-07-18T08:59:24Z INFO  collector::execute] run_rustc with incremental=false, profile=Debug, scenario=Some(Full), patch=None
[2022-07-18T08:59:25Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpseBwzm#tuple-stress@0.1.0" "--" "--wrap-rustc-with" "Eprintln"
[2022-07-18T08:59:30Z INFO  collector::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrFull), patch=None
[2022-07-18T08:59:30Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpseBwzm#tuple-stress@0.1.0" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpseBwzm/incremental-state"
[2022-07-18T08:59:36Z INFO  collector::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrUnchanged), patch=None
[2022-07-18T08:59:36Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpseBwzm#tuple-stress@0.1.0" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpseBwzm/incremental-state"
[2022-07-18T08:59:38Z DEBUG collector::execute] applying new row to "/tmp/.tmpseBwzm"
[2022-07-18T08:59:38Z INFO  collector::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("new row"), path: "0-new-row.patch" })
[2022-07-18T08:59:38Z INFO  collector::execute] run_rustc with incremental=true, profile=Debug, scenario=Some(IncrPatched), patch=Some(Patch { index: 0, name: PatchName("new row"), path: "0-new-row.patch" })
[2022-07-18T08:59:38Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmpseBwzm#tuple-stress@0.1.0" "--" "--wrap-rustc-with" "Eprintln" "-C" "incremental=/tmp/.tmpseBwzm/incremental-state"
[2022-07-18T08:59:44Z DEBUG collector::execute] Benchmark iteration 1/1
[2022-07-18T08:59:44Z INFO  collector::execute] run_rustc with incremental=false, profile=Opt, scenario=Some(Full), patch=None
[2022-07-18T08:59:44Z DEBUG collector::execute] "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp2HU2GD#tuple-stress@0.1.0" "--release" "--" "--wrap-rustc-with" "Eprintln"
[2022-07-18T08:59:50Z INFO  collector::execute] run_rustc with incremental=true, profile=Opt, scenario=Some(IncrFull), patch=None
---
   Compiling termcolor v1.1.2
[RUSTC-TIMING] unicode_xid test:false 0.087
   Compiling rustc_fs_util v0.0.0 (/checkout/compiler/rustc_fs_util)
[RUSTC-TIMING] arrayvec test:false 0.240
warning: rustc_graphviz.e22fbdc2-cgu.13: no profile data available for function _RNvXs2_NtCs7zzAukxSjOW_4core7convertNtNtNtB7_5alloc6layout11LayoutErrorINtB5_4FromBy_E4fromCsi3KMxKWQNTG_14rustc_graphviz Hash = 742261418966908927
   Compiling either v1.6.0
   Compiling either v1.6.0
warning: rustc_graphviz.e22fbdc2-cgu.10: no profile data available for function _RNvMNtNtNtCs7zzAukxSjOW_4core4iter8adapters3revINtB2_3RevINtNtNtB8_3ops5range5RangejEE3newCsi3KMxKWQNTG_14rustc_graphviz Hash = 742261418966908927
[RUSTC-TIMING] self_cell test:false 0.074
   Compiling datafrog v2.0.1
   Compiling datafrog v2.0.1
warning: rustc_fs_util.f09840a9-cgu.2: no profile data available for function _RNvXs2_NtCs7zzAukxSjOW_4core7convertNtNtNtB7_5alloc6layout11LayoutErrorINtB5_4FromBy_E4fromCs7eAO1nHmi8f_13rustc_fs_util Hash = 742261418966908927
[RUSTC-TIMING] build_script_build test:false 0.280
   Compiling memchr v2.4.1
[RUSTC-TIMING] itoa test:false 0.139
   Compiling tinyvec v0.3.4
---
   Compiling measureme v10.0.0
[RUSTC-TIMING] jobserver test:false 1.194
[RUSTC-TIMING] stacker test:false 0.292
   Compiling regex v1.5.5
warning: rustc_llvm.ef964fa5-cgu.3: no profile data available for function _RNvXs2_NtCs7zzAukxSjOW_4core7convertNtNtNtB7_5alloc6layout11LayoutErrorINtB5_4FromBy_E4fromCs6KA1KZRMfSz_10rustc_llvm Hash = 742261418966908927
[RUSTC-TIMING] generic_array test:false 0.838
   Compiling rand v0.8.5
[RUSTC-TIMING] rustc_parse_format test:false 1.033
   Compiling matchers v0.1.0
---
[RUSTC-TIMING] aho_corasick test:false 2.255
   Compiling object v0.28.4
[RUSTC-TIMING] rand test:false 1.261
   Compiling object v0.29.0
warning: rustc_serialize.ef905287-cgu.7: no profile data available for function _RINvNtCs7zzAukxSjOW_4core3ptr13drop_in_placeRjECsgIEggV8QgXX_15rustc_serialize Hash = 742261418966908927

warning: rustc_serialize.ef905287-cgu.7: no profile data available for function _RINvNtCs7zzAukxSjOW_4core9panicking13assert_failedjjECsgIEggV8QgXX_15rustc_serialize Hash = 742261418966908927

warning: rustc_serialize.ef905287-cgu.9: no profile data available for function _RNvXsV_NtCs7zzAukxSjOW_4core3fmtRjNtB5_5Debug3fmtCsgIEggV8QgXX_15rustc_serialize Hash = 1124680650125156080
[RUSTC-TIMING] rustc_serialize test:false 0.669
warning: `rustc_serialize` (lib) generated 3 warnings
[RUSTC-TIMING] tempfile test:false 0.940
[RUSTC-TIMING] sha2 test:false 1.254
---
[RUSTC-TIMING] rustc_ast_lowering test:false 1.174
error: unresolved link to `T`
    --> compiler/rustc_middle/src/ty/layout.rs:3491:24
     |
3491 | /// * uninit invalid &[T] where T has align 1 (only inside arrays)
     |                        ^ no item named `T` in scope
     |
     = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
     = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
error: could not document `rustc_middle`

Caused by:
Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustdoc --edition=2021 --crate-type lib --crate-name rustc_middle compiler/rustc_middle/src/lib.rs --target x86_64-unknown-linux-gnu -o /checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/doc -Zunstable-options --check-cfg 'values(feature, "rustc-rayon", "rustc-rayon-core", "rustc_use_parallel_compiler")' --check-cfg 'names()' --check-cfg 'values()' --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat -C metadata=4056c07e18b96fa2 -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/release/deps --extern bitflags=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/libbitflags-b08786b735ebc83e.rmeta --extern chalk_ir=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/libchalk_ir-b7d5e9cc4212a6a4.rmeta --extern either=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/libeither-c338135ee386fd69.rmeta --extern gsgdt=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/libgsgdt-18ca57f6c1c877e0.rmeta --extern polonius_engine=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/libpolonius_engine-498d962f9b7b1bd1.rmeta --extern rand=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librand-7b6b181eb887c366.rmeta --extern rand_xoshiro=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librand_xoshiro-9b11bcaa8d4cc5a8.rmeta --extern rustc_apfloat=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_apfloat-a40e8f2b6ed34d4d.rmeta --extern rustc_arena=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_arena-46e5693d0a8f70ca.rmeta --extern rustc_ast=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_ast-11872d71b90facab.rmeta --extern rustc_attr=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_attr-3dc5cb232e2aeddf.rmeta --extern rustc_data_structures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_data_structures-6cc9efdf702efa25.rmeta --extern rustc_errors=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_errors-3fa4b7305376e92a.rmeta --extern rustc_feature=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_feature-4782592e27cfed10.rmeta --extern rustc_graphviz=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_graphviz-7c2234d9aef98622.rmeta --extern rustc_hir=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_hir-79277f27b8ea3fac.rmeta --extern rustc_index=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_index-f77118641bc8969b.rmeta --extern rustc_macros=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/release/deps/librustc_macros-3402b970e104daab.so --extern rustc_query_system=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_query_system-f141b77a99f81c51.rmeta --extern rustc_serialize=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_serialize-699147027d40864b.rmeta --extern rustc_session=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_session-c14258e51a851045.rmeta --extern rustc_span=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_span-11b12bd83e996fca.rmeta --extern rustc_target=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_target-a9e3cbc0d109f75c.rmeta --extern rustc_type_ir=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_type_ir-bc57348614b023cf.rmeta --extern smallvec=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/libsmallvec-9a0501c30d2a7b13.rmeta --extern tracing=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-rustc/x86_64-unknown-linux-gnu/release/deps/libtracing-eef4fab2e6aecb3e.rmeta --extern-html-root-url 'bitflags=https://docs.rs/bitflags/1.2.1/' --extern-html-root-url 'chalk_ir=https://docs.rs/chalk-ir/0.80.0/' --extern-html-root-url 'either=https://docs.rs/either/1.6.0/' --extern-html-root-url 'gsgdt=https://docs.rs/gsgdt/0.1.2/' --extern-html-root-url 'polonius_engine=https://docs.rs/polonius-engine/0.13.0/' --extern-html-root-url 'rand=https://docs.rs/rand/0.8.5/' --extern-html-root-url 'rand_xoshiro=https://docs.rs/rand_xoshiro/0.6.0/' --extern-html-root-url 'smallvec=https://docs.rs/smallvec/1.8.1/' --extern-html-root-url 'tracing=https://docs.rs/tracing/0.1.29/' -Zunstable-options -Csymbol-mangling-version=v0 -Zunstable-options '--check-cfg=values(bootstrap)' '--check-cfg=values(parallel_compiler)' '--check-cfg=values(no_btreemap_remove_entry)' '--check-cfg=values(crossbeam_loom)' '--check-cfg=values(span_locations)' -Clink-arg=-fuse-ld=lld -Clink-arg=-Wl,--threads=1 -Dwarnings '-Wrustdoc::invalid_codeblock_attributes' --crate-version '1.64.0-nightly
  (af445596a
  2022-07-18)' --document-private-items '-Arustdoc::private-intra-doc-links' --enable-index-page -Zunstable-options -Znormalize-docs --show-type-layout --generate-link-to-definition --extern-html-root-url 'ena=https://docs.rs/ena/latest/'` (exit status: 1)
[RUSTC-TIMING] rustc_metadata test:false 1.558
[RUSTC-TIMING] rustc_infer test:false 3.150
[RUSTC-TIMING] rustc_query_impl test:false 6.035
Build completed unsuccessfully in 0:22:47
```

</details>

oli-obk commented:
@bors try

bors commented:
:hourglass: Trying commit 6ca481ecf3ca04e913712af16e07f0fc9ba91957 with merge e468a8b2c03d14f4cc595ff11f731cc44d6af743...
<!-- homu: {"type":"TryBuildStarted","head_sha":"6ca481ecf3ca04e913712af16e07f0fc9ba91957","merge_sha":"e468a8b2c03d14f4cc595ff11f731cc44d6af743"} -->

RalfJung commented:
> I did mostly follow what you described, but also figured the const eval checks for mem::zeroed would be fine since not many people misuse that

Well, except for old crossbeam...

RalfJung commented on compiler/rustc_middle/src/ty/layout.rs:
```diff
@@ -3482,3 +3482,86 @@ fn make_thin_self_ptr<'tcx>(
+        match &this.fields {
+            FieldsShape::Primitive | FieldsShape::Union { .. } => {}
+            FieldsShape::Array { count, .. } => {
+                if *count > 0 && !might_permit_raw_init_inner(this.field(cx, 0), cx, true) {
```
```suggestion
                if *count > 0 && !might_permit_raw_init_inner(this.field(cx, 0), cx, /*inside_array*/ true) {
```

RalfJung commented on compiler/rustc_middle/src/ty/layout.rs:
```diff
@@ -3482,3 +3482,86 @@ fn make_thin_self_ptr<'tcx>(
 }
+
+/// Determines if this type permits "raw" initialization by just transmuting some
+/// uninitialized memory into an instance of `T`.
```
"raw_init" was picked as a name to generalize over uninit and zero-init. So once this is uninit-only I think it should be renamed to "might_permit_uninit" or so.

But I would also hold off on that part of the change for now. Old crossbeam will not like it.

5225225 replied:
So we *should* be looking inside arrays for uninit checks, but not for zeroing checks? Since old crosssbeam makes what is effectively a zeroed `[fn(); N]`

RalfJung replied:
I think this PR should be about uninit only, yes. There we have a plan that avoids LLVM UB.

For crossbeam, no such plan can exist. Old crossbeam unambiguously has LLVM UB and we cannot work around this since `zeroed` is pretty clear about what the memory it returns will look like. So that is a separate discussion IMO. (FWIW we totally could return 0x01-filled memory when we *know* the code caused UB, since we can do anything then. Not sure if we want to.^^)

5225225 replied:
There are a *few* breakages from old crossbeam making uninit values, but from the crater run, it's a few (2 github regressions and `anterofit 0.2.0` and `constellation 0.2.0`), both having minimal download count. So yeah, ignoring zeroed failures inside arrays looks like it'll cut down on nearly all the crossbeam issues.

RalfJung commented:
Is the goal of this PR that, assuming https://github.com/rust-lang/rust/pull/99182 lands, we never have LLVM UB from `mem::uninit`? If yes then the `may_allow_raw_init` function should be documented and commented accordingly. (But this can wait until after crater, when we are sure we can land this.)

RalfJung commented on compiler/rustc_middle/src/ty/layout.rs:
```diff
@@ -3482,3 +3482,86 @@ fn make_thin_self_ptr<'tcx>(
+    where
+        C: HasDataLayout + ty::layout::HasParamEnv<'tcx> + ty::layout::HasTyCtxt<'tcx>,
+    {
+        if inside_array {
```
This new block of code here needs a ton of comments explaining why any of this makes any sense, and what even is the specification that is being implemented here.

5225225 replied:
Okay, explained as much as I could and left a link to this PR.

Commit:
Still do the type walking for zero init checks

Commit:
We ignore all references to size 0 align 1 data.

Commit:
Reword explanation for ignoring ZST 1-aligned references

5225225 commented on compiler/rustc_middle/src/ty/layout.rs:
```diff
@@ -3484,42 +3484,76 @@ fn make_thin_self_ptr<'tcx>(
-                    if let Ok(l) = cx.tcx().layout_of(penv) {
-                        return l.layout.align().abi == Align::ONE;
+            match init_kind {
+                // FIXME(#66151) We need to ignore uninit slice references with an alignment of 1
```
Okay, reworded. Added the mention of panicking on uninit if LLVM UB in the doc, and said that we require a 1-aligned 0-size value.

Kept the note of "this is to not break hyper" though, that seemed like useful background information.

RalfJung commented on compiler/rustc_middle/src/ty/layout.rs:
```diff
@@ -3482,3 +3482,143 @@ fn make_thin_self_ptr<'tcx>(
+                        let penv = ty::ParamEnv::reveal_all().and(*inner);
+                        if let Ok(l) = cx.tcx().layout_of(penv) {
+                            return l.layout.align().abi == Align::ONE
+                                && l.layout.size() == Size::ZERO;
```
Ah, I was too quick here... this does not work properly for `&dyn Trait`, we would still have LLVM UB there since the vtable is marked as 8-aligned and as dereferenceable for 24 bytes.

Our `FieldsShape::Arbitrary` branch below should catch that, but it means this early-return here may be wrong. So we probably do need to go back to what you had, where `[u8]` and `str` are special-cased, or we need to at least compute the type of the unsized tail and bail if it is a `dyn Trait`.

Please add a test ensuring that an uninit `[&dyn Send; 1]` correctly panics.

5225225 replied:
Ah yeah, that did indeed incorrectly not panic there. I reverted it back to a `[T]` where align_of(T) == 1, or a `str` check.

Also added tests that `&dyn Trait` and `*const dyn Trait` both panic when uninitialized inside arrays.

Commit:
Only panic for refs to slices and str, not dyn trait

Commit:
Adjust justification comment

RalfJung commented on src/test/ui/intrinsics/panic-uninitialized-zeroed.rs:
```diff
@@ -284,25 +284,23 @@ fn main() {
-            "attempted to zero-initialize type `LR_NonZero`, which is invalid"
+        test_panic_msg(
+            || mem::uninitialized::<(&[u8], &str, &())>(),
+            "attempted to leave type `(&[u8], &str, &())` uninitialized, which is invalid"
```
This one we could accept, if we wanted, by accepting all references to *sized* align-1-ZST... doesn't seem very important though.

RalfJung replied:
Though allowing `&[u8]` and rejecting `&()` feels rather arbitrary so we should probably allow both then?

5225225 replied:
Could do, but I figure since not many people are actually making a reference to a ZST, no harm in forbidding it now.

The changes here *are* rather arbitrary in their nature, we're only not panicking on the bits that we can't panic on just yet, while making the checks stricter where we can.

Making an uninit `[bool; 1]` will start to panic after this change, even if that's still not LLVM UB if we 0x01 fill, because there's not really a widely used crate that does that (inb4 some crate started doing that since the last crater run :p).

Hyper doesn't make uninit sized ZSTs, so I don't think we need to allow them here, even if it is somewhat inconsistent to forbid them.

I'm fine with ignoring &() as well, I just don't feel it's needed.

RalfJung commented on compiler/rustc_middle/src/ty/layout.rs:
```diff
@@ -3482,3 +3482,154 @@ fn make_thin_self_ptr<'tcx>(
+                // which acts much like a [u8].
+                //
+                // We *do* need to panic for &dyn Trait, even though the layout of dyn Trait is
+                // size 0 align 1, because &dyn Trait holds a reference to a non-zero sized type,
+                // which also must be aligned.
```
```suggestion
                // size 0 align 1, because &dyn Trait holds a reference to the vtable, which
                // must be aligned and dereferenceable.
```

Commit:
Change comment to mention vtable

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/7435026720?check_suite_focus=true)
Build commit: 138bf08760c1608a8f13fcc866a7f1c5d24bc6ab (`138bf08760c1608a8f13fcc866a7f1c5d24bc6ab`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/7435026720?check_suite_focus=true"},"merge_sha":"138bf08760c1608a8f13fcc866a7f1c5d24bc6ab"} -->

RalfJung commented:
@craterbot test

jyn514 commented:
@craterbot retry-report

Mark-Simulacrum commented:
@craterbot retry-report

Need to follow up on these, we should be retrying S3 requests...

Mark-Simulacrum commented:
@craterbot retry-report

Need to follow up on these, we should be retrying S3 requests...

5225225 commented:
> Need to follow up on these, we should be retrying S3 requests...

I checked, and the default attempt count is 3 (1 initial attempt, 2 retries, a second or so per retry).

But if we're doing a thing that has 500k+ requests, maybe that's not enough?

Mark-Simulacrum commented:
> But if we're doing a thing that has 500k+ requests, maybe that's not enough?

Hm, well, I wouldn't expect us to need to bump the retry functionality beyond the defaults in the SDK -- and the errors seem to be relatively new. I've bumped us to a more recent version of the SDK now in case there was some problem with what was retried before that might've been fixed since then. Our upload pattern is pretty benign (no concurrency, relatively slow), and generally S3 should cope with that just fine.

The archive generation error seems to have plausibly been since we were uploading the (fairly large) object in one PUT request, rather than following the recommended guideline of multipart uploads; after switching the code to multipart uploads the server is past that part of report generation and is now uploading logs.

I'm inclined to leave things as-is and if this happens again I'll start tracking it on a Crater issue and eventually investigate with S3.

Mark-Simulacrum commented:
> > Need to follow up on these, we should be retrying S3 requests...
> 
> I checked, and the default attempt count is 3 (1 initial attempt, 2 retries, a second or so per retry).
> 
> But if we're doing a thing that has 500k+ requests, maybe that's not enough?

Based on some digging, my current working theory is that after we switched to the new (but currently beta/unstable) Rust SDK rather than rusoto, we were just not retrying any of the requests -- see https://github.com/rust-lang/crater/pull/658 for the code change necessary to fix this, I hope. (This is with the caveat that we *do* retry archive uploads, since that's handled in crater today, though probably doesn't *have* to be).

I'm not sure why previous PRs didn't see quite as high an error rate -- maybe some slight change in our upload pattern or a change in S3's behavior made this more noisy.

There's an upload already going which will likely fail, but once that does I'll go ahead and retry once the new code with retries enabled actually lands.

5225225 commented:
Huh, it's been a while, is the upload stuck? I wouldn't expect an upload to take longer than, say, an hour or two? I think that's generally how long it took before. Status *is* generating report, though.

5225225 commented:
Will go through these tomorrow. I have a local `.md` file with em at the moment that I'm editing (just the crates.io regressions, it seems like every time we did one of these we largely ignored the github ones?).

@rustbot author

5225225 commented:
I'll use this comment below to track where I'm at. I went through all the crates.io regressions and noted them down here. Lot of them are just flaky tests, a lot of the regresions are some data structure (`smallvec`, `elastic-array(-plus)`) inappropriately making a slot using `mem::uninitalized`.

---

# real

* actix-sled-session: smallvec 0.6.10, ActorWaitItem (`[Box<T>; 2]`)\
  actix 0.8.3 -> smallvec 0.6.10\
  fixed by `cargo update`
  
* dot-http: smallvec 0.6.10, `[(u8, char); 4]`\
  reqwest 0.10.0 -> url 2.1.0 -> idna 0.2.0 -> unicode normalization 0.1.8 -> smallvec 0.6.10\
  `cargo update` causes build error

* double-checked-cell: crossbeam 0.2.12, Box inside enum
  scoped-pool-1.0.0 -> crossbeam-0.2.12\
  not fixed by `cargo update`

* dubs-cloudflare: smallvec 0.6.10 `[(u8, char); 4]`
  reqwest 0.9.20 -> url 1.7.2 -> idna 0.1.5 -> unicode-normalization 0.1.8 -> smallvec 0.6.10\
  fixed by `cargo update`

* etcommon-block: elastic-array-plus 0.10.0
* etcommon-block-core: elastic-array-plus 0.10.0
* etcommon-rlp: elastic-array-plus 0.10.0
* etcommon-trie: elastic-array-plus 0.10.0
* etcommon-trie-test: elastic-array-plus 0.10.0
* ethereum-rlp: elastic-array-plus 0.9.1
* ethereum-trie: elastic-array-plus 0.9.1
* ethereum-trie-memory: elastic-array-plus 0.9.1
* ethereum-trie-test: elastic-array-plus 0.9.1
* eventador: lockfree 0.5.1 (creates `Option<Box<dyn Trait>>`)
* git-req: smallvec 0.6.9, `[(u8, char); 4]`
* hash-rings: smallvec 0.6.10, `[sieve::Item; 1]`
* hash-trie: own code, `[Enum; 1]`
* hexagon: own code, `[Cell<Enum> ;N]`
* jsonbox: smallvec 0.6.10, `[(u8, char); 4]`
* keypad: own code, array containing non-ZST reference
* lazy-init: crossbeam, enum
* limitation-actix-middleware: smallvec 0.6.10, `[(u8, char); 4]`
* limitation-proxy: smallvec 0.6.10, `[(u8, char); 4]`
* makods: own code, `Option<Box<T>>`
* mostinefficientsha: arrayvec 0.3.25, `[Rc<T>; 32]`
* oftlisp: smallvec 0.4.5, `[char; 4]`
* rocketchat_client_rs: smallvec 0.6.10, `[(u8, char); 4]`
* rudy: own code, array containing non-ZST reference
* rusteam: smallvec 0.6.10, `[(u8, char); 4]`
* signal-simple: own code, sigaction
* solana-validator-info: smallvec 0.6.10, `[(u8, char); 4]`
* sputnikvm-rlp: elastic-array 0.7.0, enum inside array
* sputnikvm-stateful: elastic-array-plus 0.10.0, enum inside array
* structdoc: smallvec 0.6.10, `[(u8, char); 4]`
* tdns-cli: smallvec 0.6.10, `[(u8, char); 4]`
* teleterm: tokio-sync 0.1.7, enum inside array
* tempest: smallvec-0.6.10, Box inside array
* term-handler: signal-simple-0.1.1, sigaction
* zotero: smallvec-0.6.10, `[(u8, char); 4]`


# flaky

boolean_expression btctipserver catboost2 catboost2-sys combine-language
completable_future emlop git-snapshot hawktracer-converter heca hopscotch ijson
impulse_response intervalier jwt-cli kai kzg lares ledger-transport-zemu
lf-queue lightbird linemux loam metrics_server mock-rs nn openexr-sys pickup
psyche-serde quanta revenq rt-thread rusty_planner shack similari
simple-irc-server sslhash standard-dist stringsort strip-ansi-escapes
tcp-stream-echo term-handler testdir threadfin tokio-stream-ext
tower-fault-injector uuid7 vebtrees

# segfaults

esr gfwx ggrep core_cbc_casper keyphrase nude potatonet qjsonrs s3mon simimgrs
tcp-clone tokio-fs tokio-threadpool

# unknown

control-code: too long, need to check out
filemagic: looks random? "stream doesn't contain UTF-8"?
multiqueue: double panic
pendulum: double panic
specs-mirror: double panic

5225225 commented:
Okay, went through all of them. I didn't create an issue for the ones that I couldn't reproduce / worked fine with a fresh lockfile, but I reported most of them upstream.

Still need to edit the PR description before this is merged to describe what was done but I figure I might as well mark this ready.

@rustbot ready

5225225 commented:
https://github.com/rust-lang/rust/issues/98862#issuecomment-1209531725 gave a go-ahead to introduce a FCW. Might be smart to do that before this, so people have some warning period. (Though that would be for all uses of mem::uninitialized, and not just the ones we're panicking on here)

5225225 commented:
Gonna mark this as blocked on #100342. Once/if the FCW goes in and people get a chance to bump dependencies, it'll be worth revisiting this, but as stated, I think it'd be unwise to get this in first.

@rustbot label +S-blocked -S-waiting-on-review

RalfJung commented:
https://github.com/rust-lang/rust/pull/101061 is an alternative to this PR that relaxes the panics in light of `mem::uninitiailized` now filling its return value with 0x01. That PR + just recursing through arrays should still work with hyper, and keeps the logic a lot simpler than this PR.

RalfJung commented:
Actually, this makes me wonder -- is there even LLVM UB from `mem::uninitialized` on arrays? We never use scalar layout for arrays, so I don't think we emit any metadata that would cause LLVM UB from an uninitialized unused array. In that sense there seems to be no good reason to break old crossbeam.

Maybe stopping at arrays is actually fine if all we are aiming for is avoid LLVM UB -- and the only thing we still should strengthen are the checks for `Variants::Single`?

5225225 commented:
Closing this seeing that #101061 was implemented, this is not really needed anymore
